File: extension/handlers/execute_script.js
'use strict';

/**
 * execute_script — escape hatch for one-off automation.
 *
 * Two execution targets:
 *   "main"  — runs `(async () => { <code> })()` inside the extension's
 *             main process. Has access to Editor.* APIs (assetdb, log, etc.)
 *             but NOT to cc.director / scene nodes.
 *   "scene" — forwards the snippet to scene-script.js, where cc.director
 *             and the live scene are available.
 *
 * The result is whatever the snippet returns (must be JSON-serializable).
 *
 * SAFETY NOTE: this is a deliberately powerful primitive — the Python
 * wrapper marks it as requiring explicit user permission. Don't use it
 * for things where a typed handler exists.
 */

const Util = require('util');

function _exec_main(code) {
    if (typeof code !== 'string' || !code.trim()) {
        throw new Error('execute_script: `code` must be a non-empty string');
    }
    // eslint-disable-next-line no-new-func
    const wrapped = new Function('Editor', 'require', '__dirname',
        '"use strict"; return (async () => {\n' + code + '\n})();');
    return Promise.resolve(wrapped(typeof Editor !== 'undefined' ? Editor : undefined, require, __dirname))
        .then((value) => ({ ok: true, value: _safeSerialize(value) }))
        .catch((err) => { throw err; });
}

function _exec_scene(code) {
    if (typeof code !== 'string' || !code.trim()) {
        throw new Error('execute_script: `code` must be a non-empty string');
    }
    return new Promise((resolve, reject) => {
        Editor.Scene.callSceneScript('cocos-mcp', 'mcp:exec', { code }, (err, data) => {
            if (err) { reject(err instanceof Error ? err : new Error(String(err))); return; }
            resolve({ ok: true, value: data });
        });
    });
}

function _safeSerialize(v) {
    if (v === undefined) return null;
    try {
        // round-trip to make sure the result is JSON-clean
        return JSON.parse(JSON.stringify(v));
    } catch (e) {
        return Util.inspect(v, { depth: 4, breakLength: 120 });
    }
}

async function handle(params, ctx) {
    const target = (params && params.target) || 'main';
    const code = params && params.code;
    if (target === 'main') return await _exec_main(code);
    if (target === 'scene') return await _exec_scene(code);
    throw new Error('execute_script: unknown target ' + target + ' (valid: main, scene)');
}

module.exports = { name: 'execute_script', handle };

File: server/src/services/tools/execute_script.py
"""execute_script tool — escape hatch that runs arbitrary JS in the editor."""

from __future__ import annotations

from typing import Annotated, Any, Literal

from fastmcp import Context

from services.registry import cocos_mcp_tool
from services.tools._common import call_bridge


@cocos_mcp_tool(
    description=(
        "EXECUTE ARBITRARY JAVASCRIPT inside the Cocos Creator editor. "
        "This is a powerful primitive — prefer the typed tools when they "
        "cover the use case, and ask the user before calling this for "
        "anything destructive.\n\n"
        "Targets:\n"
        "  main   - runs in the extension's main process. Has Editor.* APIs "
        "(assetdb, log, project paths, ...) but NOT cc.director / live scene.\n"
        "  scene  - runs in the scene context. Has cc, cc.director, the active "
        "scene as `scene`. Use this to walk/edit nodes, components, prefabs.\n\n"
        "Snippets are wrapped as `(async () => { <code> })()`. The return "
        "value of that async IIFE is sent back as the tool's `data.value` "
        "(must be JSON-serializable; otherwise we coerce to a string)."
    ),
)
async def execute_script(
    ctx: Context,
    code: Annotated[str, "JavaScript snippet to execute."],
    target: Annotated[
        Literal["main", "scene"],
        "Where to run the snippet. Default 'main'.",
    ] = "main",
) -> dict[str, Any]:
    return await call_bridge("execute_script", {"code": code, "target": target},
                             timeout=60)
